src/SonarScanner.MSBuild.Shim/JsonPropertiesWriter.cs

```diff
@@ -18,16 +18,275 @@
+    /// </summary>
+    public void WriteGlobalSettings(AnalysisProperties properties)
+    {
+        if (properties is null)
+        {
+            throw new ArgumentNullException(nameof(properties));
+        }
+
+        foreach (var setting in properties)
+        {
+            // We should no longer pass the sonar.verbose=true parameter to the scanner CLI.
+            // See: https://github.com/SonarSource/sonar-scanner-msbuild/issues/543
+            if (setting.Id != SonarProperties.Verbose)
+            {
+                AppendKeyValue(setting.Id, setting.Value);
+            }
+        }
+    }
+
+    public void WriteSonarProjectInfo(DirectoryInfo projectBaseDir)
+    {
+        AppendKeyValue(SonarProperties.ProjectKey, config.SonarProjectKey);
+        AppendKeyValueIfNotEmpty(SonarProperties.ProjectName, config.SonarProjectName);
+        AppendKeyValueIfNotEmpty(SonarProperties.ProjectVersion, config.SonarProjectVersion);
+        AppendKeyValue(SonarProperties.WorkingDirectory, Path.Combine(config.SonarOutputDir, ".sonar"));
+        AppendKeyValue(SonarProperties.ProjectBaseDir, projectBaseDir.FullName);
+        AppendKeyValue(SonarProperties.PullRequestCacheBasePath, config.GetConfigValue(SonarProperties.PullRequestCacheBasePath, null));
+    }
+
+    public void WriteSharedFiles(AnalysisFiles analysisFiles)
+    {
+        if (analysisFiles.Sources.Count > 0)
+        {
+            AppendKeyValue("sonar", "sources", analysisFiles.Sources);
+        }
+        if (analysisFiles.Tests.Count > 0)
+        {
+            AppendKeyValue("sonar", "tests", analysisFiles.Tests);
+        }
+    }
+
+    internal /* for testing purposes */ string EncodeAsMultiValueProperty(IEnumerable<string> paths)
+    {
+        var multiValuesPropertySeparator = @$",\{Environment.NewLine}";
+
+        if (Version.TryParse(config.SonarQubeVersion, out var sonarqubeVersion) && sonarqubeVersion.CompareTo(new Version(6, 5)) >= 0)
+        {
+            return string.Join(multiValuesPropertySeparator, paths.Select(x => $"\"{x.Replace("\"", "\"\"")}\""));
+        }
+        else
+        {
+            var invalidPaths = paths.Where(InvalidPathPredicate);
+            if (invalidPaths.Any())
+            {
+                logger.LogWarning(Resources.WARN_InvalidCharacterInPaths, string.Join(", ", invalidPaths));
+            }
+
+            return string.Join(multiValuesPropertySeparator, paths.Where(x => !InvalidPathPredicate(x)));
+        }
+    }
+
+    private void AppendKeyValue(string keyPrefix, string keySuffix, IEnumerable<FileInfo> paths) =>
+        AppendKeyValue(keyPrefix, keySuffix, paths.Select(x => x.FullName));
+
+    private void AppendKeyValue(string keyPrefix, string keySuffix, IEnumerable<string> paths) =>
+        jsonArray.Add(new JObject
+            {
+                { "key", $"{keyPrefix}.{keySuffix}" },
+                { "value", new JRaw($"\"{EncodeAsMultiValueProperty(paths)}\"") }
```

Comment:
With JRaw the special characters seem to work.
Instead of "dir//file,//r/nAnother//Path" it stores in the the JArray as
"dir/file,/
Another/Path"

Reply:
This implementation is wrong and I'll have to rework it. It currently produces something like
```
value": "File.cs,\\r\n你好.cs"
```
Problems:
* \r doesn't seem to be liked by the Scanner engine, 
* There's unescaped `\` after `,`, making it `backslash letter-r carride-return` sequence
* It does not confirm to the parsing logic

Actual parsing logic is here
https://github.com/SonarSource/sonarqube/blob/master/sonar-plugin-api-impl/src/main/java/org/sonar/api/config/internal/MultivalueProperty.java#L46

The expected format is simply comma-separated CSV of values
```
value": "File.cs,你好.cs"
```

I have a working prototype with manually crafted JSON input and manual invocation of the Scanner engine that verifies this claim.